### PR TITLE
KEP-4622: Add docs for TopologyManager policy option for MaxAllowableNUMANodes

### PR DESCRIPTION
**Pull request by LaurentGoderre**
Add documentation for KEP-4622

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 1]
```diff
@@ -222,17 +222,26 @@ You will still have to enable each option using the `TopologyManagerPolicyOption
...
+  You can enable this option by adding `prefer-closest-numa-nodes=true` to the Topology Manager policy options.
+  By default, without this option, Topology Manager aligns resources on either a single NUMA node or
+  the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However,
+  the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.
+  This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems,
+  and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
+  Topology Manager decides to align resources on non-adjacent NUMA nodes.
+
+* `max-allowable-numa-nodes` (alpha; `TopologyManagerPolicyOptions` and `TopologyManagerPolicyBetaOptions` feature gates have to be enabled)
+  The `max-allowable-numa-nodes` policy option is alpha in Kubernetes {{< skew currentVersion >}}.
```
Comment (cyclinder) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:238]: `max-allowable-numa-nodes` is in beta stage.

Comment (LaurentGoderre) [reply content/en/docs/tasks/administer-cluster/topology-manager.md]: Done!

**Comment by Princesso** [comment]
Hello @LaurentGoderre  :wave: please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday July 16th 2024 18:00 PST. When it is ready for review, please remove the Draft/WIP status. Thank you!

**Comment by LaurentGoderre** [comment]
@cyclinder are you ok with the PR being ready to review?

**Comment by LaurentGoderre** [comment]
@cyclinder also can you add a link to this PR to https://github.com/kubernetes/enhancements/issues/4622

**Comment by cyclinder** [comment]
@LaurentGoderre I'm happy with the PR being ready to review, and let me add this link now.

**Comment by Princesso** [comment]
Hello @cyclinder  :wave:! v1.31 Doc Lead here. Just checking in as we approach Docs Freeze on Tuesday July 30th 18:00 PDT. This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as lgtm and approve labels applied, without any unaddressed comments or concerns from SIG Docs. The status of this enhancement is marked as at risk for docs freeze. Thank you!

**Comment by kannon92** [comment]
/cc @ffromani 

Can you give a technical review as our expert on TopologyManager?

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 2]
```diff
@@ -222,17 +222,26 @@ You will still have to enable each option using the `TopologyManagerPolicyOption
...
+  policies will favor sets of NUMA nodes with shorter distance between them when making admission decisions.
+  You can enable this option by adding `prefer-closest-numa-nodes=true` to the Topology Manager policy options.
+  By default, without this option, Topology Manager aligns resources on either a single NUMA node or
+  the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However,
+  the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.
+  This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems,
+  and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
+  Topology Manager decides to align resources on non-adjacent NUMA nodes.
+
+* `max-allowable-numa-nodes` (beta; `TopologyManagerPolicyOptions` and `TopologyManagerPolicyBetaOptions` feature gates have to be enabled)
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:237]: ok, this is not related to this very change because I see the same text at line 225: this is not completely correct. The BetaOptions gate is always enabled by design. Please see https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L822 and https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L1200 .
IOW, the gates are currently enabled by default so no action is needed by default

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 3]
```diff
@@ -222,17 +222,26 @@ You will still have to enable each option using the `TopologyManagerPolicyOption
...
+  The `max-allowable-numa-nodes` policy option is beta in Kubernetes {{< skew currentVersion >}}.
+
+  The value of `max-allowable-numa-nodes` does not (in and of itself) affect the latency of pod admission. With the `TopologyManager` enabled,
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:240]: With the `TopologyManager` enabled -> When the Topology Manager policy is not `none`...
because the TopologyManager graduated to GA, so it's always enabled. However, setting the policy to `none` will bypass almost all its logic, rendering it a NOP

Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:240]: `The value of `max-allowable-numa-nodes` does not (in and of itself) affect the latency of pod admission.`
I'm not a native english speaker, but I'd put this detail at the end, once we described what the option do.

**Review by ffromani — COMMENTED** [review]
few clarifications needed about the wording, but overall good

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 4]
```diff
@@ -222,17 +222,26 @@ You will still have to enable each option using the `TopologyManagerPolicyOption
...
+  the time to admit a pod is tied to the number of NUMA nodes on the physical machine. With the new `max-allowable-numa-nodes` option, nodes
+  with more than 8 NUMA nodes can be alowed to run with the `TopologyManager` enabled. However, it is unknown exactly how much this will slow
```
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:242]: alowed -> allowed

Does this mean there is a soft threshold for `max-allowable-numa-nodes`?

Comment (ffromani) [reply content/en/docs/tasks/administer-cluster/topology-manager.md:242]: TL;DR: we don't know yet.

the gist of this option is that the limit was added in the first place when machines with > 8 NUMA nodes were very rare, so the cap was set "high enough" because there wasn't enough data about how the topology manager algorithm scale.

Nowadays machines with > 8 NUMA nodes are common, and the kubelet just errors out in this scenario (https://github.com/kubernetes/kubernetes/issues/124144 and https://github.com/kubernetes/kubernetes/blob/release-1.30/pkg/kubelet/cm/topologymanager/topology_manager.go#L155 - this is in the setup stage)

So: we don't know how much the admission will slow down with more NUMA nodes to deal with. There's no enough data. Users are now enabled to at least try instead of getting a runtime error (and kubelet crash).

**Comment by Princesso** [comment]
Hello @LaurentGoderre, @cyclinder,

v1.31 Doc Lead here. This PR has been marked as at risk for Doc Freeze.

The Doc Reviewers have given feedback on it, however, an update is yet to be made on it, based on feedback.

Please note that the Doc freeze deadline is tomorrow, and if the PR is not merged by tomorrow, you will need to file an exception or this enhancement will be removed from this release. Thank you!

**Comment by cyclinder** [comment]
hi @Princesso, Can you tell me where to file this exception? thanks!

**Comment by katcosgrove** [comment]
> hi @Princesso, Can you tell me where to file this exception? thanks!

The exception process is [here](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md). 

If you are unable to address this feedback and get reviews done in time, please file an exception request as soon as possible.

**Comment by cyclinder** [comment]
Yes, I can't address this feedback, @LaurentGoderre please :) I'm trying to file an exception.

**Comment by cyclinder** [comment]
I have filed an exception, see https://groups.google.com/g/kubernetes-sig-release/c/zn1JGEn6V30

**Comment by LaurentGoderre** [comment]
@ffromani @cyclinder updated the PR, can you review again?

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 5]
```diff
@@ -221,18 +221,29 @@ You can toggle groups of options on and off based upon their maturity level usin
...
-If the `prefer-closest-numa-nodes` policy option is specified, the `best-effort` and `restricted`
-policies will favor sets of NUMA nodes with shorter distance between them when making admission decisions.
-You can enable this option by adding `prefer-closest-numa-nodes=true` to the Topology Manager policy options.
-By default, without this option, Topology Manager aligns resources on either a single NUMA node or
-the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However,
-the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.
-This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems,
-and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
-Topology Manager decides to align resources on non-adjacent NUMA nodes.
+* `prefer-closest-numa-nodes` (beta, and enabled by default).
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:224]: I'd change to "beta, and visible by default".

The current unfortunately still not correct. In the context of the options, this means the feature is `visible` by default. Otherwise, you need to enable the feature gate to be able to set the option, and in turn the option default is the backward compatible one.

So it's a 2-level protection:
1. enable the option visibility (default: visible for beta, *not* visible for alpha)
2. enable the option proper

to summarize, the beta options are visible (not enabled) by default.

Sorry for any confusion caused.

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 6]
```diff
@@ -221,18 +221,29 @@ You can toggle groups of options on and off based upon their maturity level usin
...
+  policies will favor sets of NUMA nodes with shorter distance between them when making admission decisions.
+  You can enable this option by adding `prefer-closest-numa-nodes=true` to the Topology Manager policy options.
+  By default, without this option, Topology Manager aligns resources on either a single NUMA node or
+  the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However,
+  the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.
+  This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems,
+  and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
+  Topology Manager decides to align resources on non-adjacent NUMA nodes.
+
+* `max-allowable-numa-nodes` (beta, and enabled by default).
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:237]: likewise, `visible by default`

**Review by ffromani — COMMENTED** [review]
thanks! we need one last change s/enabled by default/visible by default/ and we're good to go. Details inline.

**Comment by ffromani** [comment]
@LaurentGoderre thanks, almost there. We need one final change and LGTM.

File: content/en/docs/tasks/administer-cluster/topology-manager.md (hunk 5 above)
Comment (LaurentGoderre) [reply content/en/docs/tasks/administer-cluster/topology-manager.md]: @ffromani this one I changed by accident and is part of another KEP.

**Comment by LaurentGoderre** [comment]
@ffromani do we change the wording of the other option that was already there? See https://github.com/kubernetes/website/commit/afbe795bdbe6aa08bcefcd1ce999f81fa6990599

**Comment by ffromani** [comment]
/lgtm

@LaurentGoderre thanks, simpler is better. I'd keep it this way.

**Comment by ffromani** [comment]
LGTM from sig-node perspective

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 7]
```diff
@@ -234,6 +234,17 @@ This limitation surfaces in multi-socket, as well as single-socket multi NUMA sy
 and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
 Topology Manager decides to align resources on non-adjacent NUMA nodes.
 
+* `max-allowable-numa-nodes` (beta, visible by default).
+The `max-allowable-numa-nodes` policy option is beta in Kubernetes {{< skew currentVersion >}}.
+
+With the `TopologyManager` enabled, the time to admit a pod is tied to the number of NUMA nodes on the physical machine.
+With the new `max-allowable-numa-nodes` option, nodes with more than 8 NUMA nodes can be alowed to run
```
Comment (katcosgrove) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:241]:
```suggestion
With the new `max-allowable-numa-nodes` option, nodes with more than 8 NUMA nodes can be allowed to run
```

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 8]
```diff
@@ -234,6 +234,17 @@ This limitation surfaces in multi-socket, as well as single-socket multi NUMA sy
...
+with the `TopologyManager` enabled. However, it is unknown exactly how much this will slow
+down pod admission on any given system. This feature is therefore to be used at-your-own-risk until we have
```
Comment (katcosgrove) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:243]:
```suggestion
down pod admission on any given system. This feature is therefore to be used at your own risk until we have
```

**Review by katcosgrove — COMMENTED** [review]
Grammatical nits

File: content/en/docs/tasks/administer-cluster/topology-manager.md [hunk 9]
```diff
@@ -234,6 +234,17 @@ This limitation surfaces in multi-socket, as well as single-socket multi NUMA sy
...
+a proper solution in place to reduce the state explosion that causes pod admission time to slow down as the
+number of NUMA nodes increases. The value of `max-allowable-numa-nodes` does not (in and of itself) affect the
+latency of pod admission.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:246]:
```suggestion
The time to admit a pod is tied to the number of NUMA nodes on the physical machine.
If you select the the `max-allowable-numa-nodes` policy option, nodes with more than 8 NUMA nodes can
be allowed to run with the topology enabled. The Kubernetes project only has limited data of the impact
of using the topology manager on (Kubernetes) nodes with more than 8 NUMA does. Because of that
lack of data, using this policy option is **not** recommended and is at your own risk.

Setting a value of `max-allowable-numa-nodes` does not (in and of itself) affect the
latency of pod admission, but binding a Pod to a (Kubernetes) node with many NUMA does does
have an impact.
Future, potential improvements to Kubernetes may improve Pod admission performance and the high
latency that happens as the number of NUMA nodes increases.
```

We usually [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future); this one is couched so that it aligns with the spirit of that guidance.

File: content/en/docs/tasks/administer-cluster/topology-manager.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/topology-manager.md:1]: Ideally, fix the formatting so that “_The following policy options exists:_” is followed by a _single_ unordered list.

**Review by sftim — COMMENTED** [review]
Thanks. I recommend further changes; see inline feedback with suggestion.

**Comment by sftim** [comment]
See https://github.com/kubernetes/website/pull/46870#pullrequestreview-2207754929
/lgtm cancel

File: content/en/docs/tasks/administer-cluster/topology-manager.md (hunk 7 above)
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/topology-manager.md:241]: We also (strongly) avoid “new“; we like documentation to be [timeless](https://developers.google.com/style/timeless-documentation).

I'd make an exception for a page that was entirely about an alpha feature, because the quality bar is lower for alpha docs.

**Comment by LaurentGoderre** [comment]
@sftim I had to fix the formatting of the previous block to make it consistent

**Review by sftim — COMMENTED** [review]
Thanks

/lgtm
/approve
